src/main/java/bowling/Frame.java

```diff
@@ -0,0 +1,119 @@
+package bowling;
+
+import bowling.exception.BowlingException;
+import bowling.exception.FrameOverException;
+import bowling.exception.InvalidPinQuantityException;
+
+public class Frame {
+    public static final int MAX_PINS = 10;
+    private final boolean isFinalFrame;
+    private int pinsStanding = MAX_PINS;
+    private int rollNumber = 0;
+    private boolean isOver = false;
+    private int rollOnePins = 0;
+    private int rollTwoPins = 0;
+    private FrameType frameType = FrameType.NORMAL;
+    private int score = 0;
+
+    public Frame(boolean isFinalFrame) {
+        this.isFinalFrame = isFinalFrame;
+    }
+
+    public Frame() {
+        this(false);
+    }
+
+    public void roll(int numberOfPinsKnockedDown) throws BowlingException {
+        if (isOver) {
+            throw new FrameOverException("This frame is over, wyd fam");
+        }
+
+        if (numberOfPinsKnockedDown > pinsStanding) {
+            throw new InvalidPinQuantityException("R u dizzy u knocked over  more pins than there r");
+        }
+
+        rollNumber++;
+        pinsStanding -= numberOfPinsKnockedDown;
+
+        if (isFinalFrame) {
+            rollFinalFrame(numberOfPinsKnockedDown);
+        } else {
+            rollNonFinalFrame(numberOfPinsKnockedDown);
+        }
+    }
+
+    public void rollNonFinalFrame(int numberOfPinsKnockedDown) {
+        switch (rollNumber) {
+            case 1:
+                rollOnePins = numberOfPinsKnockedDown;
+                score += rollOnePins;
+
+                if (numberOfPinsKnockedDown == MAX_PINS) {
+                    this.frameType = FrameType.STRIKE;
+                    this.isOver = true;
+                }
+
+                break;
+            case 2:
+                rollTwoPins = numberOfPinsKnockedDown;
+                score += rollTwoPins;
+
+                if (rollOnePins + rollTwoPins == MAX_PINS) {
+                    this.frameType = FrameType.SPARE;
+                }
+
+                this.isOver = true;
+                break;
+            }
```

Comment:
It is good practice to have a default that errors in case something super weird has happened in the application. A catch-all behaviour. Otherwise they will go unnoticed and you could have obscure bugs flying under the radar.